File: support/src/main/java/com/wuguangxin/ui/XinBaseActivity.java
package com.wuguangxin.ui;

import android.app.Activity;
import android.app.Dialog;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.util.SparseArray;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.View;

import com.scwang.smart.refresh.layout.SmartRefreshLayout;
import com.wuguangxin.base.FragmentTask;
import com.wuguangxin.base.LayoutManager;
import com.wuguangxin.base.LoadingStatus;
import com.wuguangxin.base.TitleBar;
import com.wuguangxin.dialog.LoadingDialog;
import com.wuguangxin.dialog.XinDialog;
import com.wuguangxin.listener.BaseInterface;
import com.wuguangxin.support.R;
import com.wuguangxin.utils.AndroidUtils;
import com.wuguangxin.utils.AnimUtil;
import com.wuguangxin.utils.Logger;
import com.wuguangxin.utils.SlidingFinishHelper;
import com.wuguangxin.utils.SoftHideKeyBoardUtil;
import com.wuguangxin.utils.StatusBarUtils;
import com.wuguangxin.utils.ToastUtils;

import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.Fragment;
import butterknife.ButterKnife;
import butterknife.Unbinder;

/**
 * Activity基类
 * Created by wuguangxin on 2015/4/1
 */
public abstract class XinBaseActivity extends AppCompatActivity implements BaseInterface {
    protected SmartRefreshLayout mRefreshLayout;
    protected LayoutManager mLayoutManager;
    protected LoadingDialog mLoadingDialog;
    protected XinDialog mDialog;
    protected Context mContext;
    private Unbinder mUnBinder;
    protected String TAG;
    protected int screenWidth;
    private boolean slidingFinish;
    private SlidingFinishHelper mSlidingFinishHelper;

    @Override
    final protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        TAG = getClass().getSimpleName();
        mContext = this;
        screenWidth = AndroidUtils.getScreenWidth(this);
        mLoadingDialog = new LoadingDialog(this); // 加载对话框
        mSlidingFinishHelper = new SlidingFinishHelper(this);

        mLayoutManager = new LayoutManager(this, R.layout.activity_base); // 布局管理器
        mLayoutManager.setContentView(getLayoutRes());
        mLayoutManager.setErrorLayoutListener(v -> onClickErrorLayout());

        setContentView(mLayoutManager.getRootLayout());
        mUnBinder = ButterKnife.bind(this, this);

        getTitleBar().setTitle(getClass().getSimpleName());
        SoftHideKeyBoardUtil.assistActivity(this);

        setImmersionStatusBar(this);

        initArguments(getIntent());
        initView(savedInstanceState);
        initListener();
    }

    /**
     * 设置沉浸式状态栏
     *
     * @param activity
     */
    public void setImmersionStatusBar(Activity activity) {
        StatusBarUtils.setImmersionStatusBar(activity, getResources().getColor(R.color.xin_titlebar_background));
        StatusBarUtils.setStatusBarFontColor(activity, false);
    }

    /**
     * 初始化一些传参
     *
     * @param intent
     */
    public void initArguments(Intent intent) {
    }

    /**
     * 是否开启滑动关闭 Activity
     */
    public void setSlidingFinish(boolean slidingFinish) {
        this.slidingFinish = slidingFinish;
    }

    @Override
    public boolean dispatchTouchEvent(MotionEvent ev) {
        if (slidingFinish) {
            return mSlidingFinishHelper.dispatchTouchEvent(ev);
        }
        return super.dispatchTouchEvent(ev);
    }

    @Override
    protected void onStart() {
        super.onStart();
        initData();
    }

    @Override
    protected void onDestroy() {
        // 关闭所有对话框dismiss
        getTitleBar().setLoadAnimVisible(false);
        dismissDialog();
        // 销毁Presenter
//        if (mPresenter != null) {
//            mPresenter.onDestroy();
//            mPresenter = null;
//        }
        if (mUnBinder != null) {
            mUnBinder.unbind();
            mUnBinder = null;
        }
        if (viewSparse != null) {
            viewSparse.clear();
            viewSparse= null;
        }
        super.onDestroy();
    }

    @Override
    public void finish() {
        super.finish();
        AnimUtil.animClose(this);
    }

    /**
     * 当点击错误界面时执行的方法。
     * （使用场景：某个界面第一次请求数据并发生网络异常时，显示错误界面，点击错误界面时，执行该方法，例如重新请求数据。）
     */
    public void onClickErrorLayout() {
        initData();
    }

    /**
     * 获取当前使用的 SmartRefreshLayout
     */
    protected SmartRefreshLayout getRefreshLayout() {
        if (mRefreshLayout == null) {
            LayoutInflater inflater = LayoutInflater.from(this);
            mRefreshLayout = (SmartRefreshLayout) inflater.inflate(R.layout.xin_def_refresh_layout, null);
        }
        return mRefreshLayout;
    }

    public LayoutManager getLayoutManager() {
        return mLayoutManager;
    }

    public TitleBar getTitleBar() {
        return getLayoutManager().getTitleBar();
    }

    /**
     * 取消透明状态栏样式
     */
    public void clearImmersionStatusBar() {
        AndroidUtils.clearImmersionStatusBar(this);
    }

    /**
     * 该方法会将传入的layoutRes布局放入到SmartRefreshLayout里，实现下拉刷新，这只是少写一层代码而已。
     *
     * @param layoutRes
     */
    public void setContentViewOnRefreshLayout(int layoutRes) {
        View view = LayoutInflater.from(mContext).inflate(layoutRes, getRefreshLayout());
        setContentView(view);
    }

//    @Override
//    public void setContentView(int layoutRes) {
//        getLayoutManager().setContentView(layoutRes);
//        mUnBinder = ButterKnife.bind(this, this);
//    }

//    @Override
//    public void setContentView(View view) {
//        getLayoutManager().setContentView(view);
//        mUnBinder = ButterKnife.bind(this, this);
//    }

    @Override
    public void setTitle(CharSequence title) {
        super.setTitle(title);
        getTitleBar().setTitle(title);
    }

    @Override
    public void setTitle(int title) {
        super.setTitle(title);
        getTitleBar().setTitle(title);
    }

    @Override
    public void setTitleColor(int color) {
        getTitleBar().getTitleView().setTextColor(color);
    }

    protected void setTitleBarVisibility(boolean visibility) {
        getTitleBar().setVisibility(visibility);
    }

    protected void setBodyVisibility(boolean visibility) {
        getLayoutManager().setBodyVisibility(visibility);
    }

    // ============================ BaseViewListener ===============================================
    @Override
    public Context getContext() {
        return mContext;
    }

    private SparseArray<View> viewSparse = new SparseArray<>();

    @Override
    public <T extends View> T findView(int id) {
        View view = viewSparse.get(id);
        if (view == null) {
            view = findViewById(id);
            viewSparse.put(id, view);
        }
        return (T) view;
    }

    @Override
    public void showToast(String text) {
        ToastUtils.showToast(getContext(), text);
    }

    @Override
    public void printLogI(String text) {
        Logger.i(getContext(), text);
    }

    @Override
    public void openActivity(Class<? extends Activity> clazz) {
        openActivity(clazz, null);
    }

    @Override
    public void openActivity(Class<? extends Activity> clazz, Bundle bundle) {
        Intent intent = new Intent(getContext(), clazz);
        intent.putExtra("bundle", bundle);
        startActivity(intent);
    }

    @Override
    public void startActivity(Intent intent) {
        // super.startActivity(intent);
        // 换成startActivityForResult打开的Activity切换动画比较好看，并且requestCode=0
        super.startActivityForResult(intent, 0);
    }

    @Override
    public void setLoadingStatus(int loadingStatus, boolean isPull, boolean isCached) {
        switch (loadingStatus) {
        case LoadingStatus.START:
            if (isPull && isCached) {
                setTitleLoadingProgressVisible(true);
            } else {
                setLoadingDialogVisible(true);
            }
            break;
        case LoadingStatus.SUCCESS:
        case LoadingStatus.FAILURE:
            // 是获取数据时，且没有缓存数据，且网获取失败，则可显示网络错误界面
            if (isPull && !isCached && loadingStatus == LoadingStatus.FAILURE) {
                getLayoutManager().setErrorLayoutVisible(true);
            }
            if (isCached) {
                setTitleLoadingProgressVisible(false);
            } else {
                setLoadingDialogVisible(false);
            }
            break;
        case LoadingStatus.CANCEL:
        case LoadingStatus.FINISH:
            setTitleLoadingProgressVisible(false);
            setLoadingDialogVisible(false);
            break;
        }
    }

    @Override
    public void setLoadingDialogVisible(boolean visible) {
        if (mLoadingDialog != null) {
            mLoadingDialog.setVisible(visible);
        }
    }

    @Override
    public void setTitleLoadingProgressVisible(boolean isStart) {
        getTitleBar().setLoadAnimVisible(isStart);
    }

    @Override
    public void dismissDialog() {
        if (mLoadingDialog != null) {
            mLoadingDialog.dismiss();
            mLoadingDialog = null;
        }
        if (mDialog != null) {
            mDialog.dismiss();
            mDialog = null;
        }
    }

    @Override
    public void dismissDialog(Dialog... dialogs) {
        for (Dialog dialog : dialogs) {
            if (dialog != null) {
                dialog.dismiss();
            }
        }
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        // 让最后打开的那个fragment执行
        Fragment topFragment = FragmentTask.getInstance().getTopTask();
        if (topFragment != null) {
            topFragment.onActivityResult(requestCode, resultCode, data);
        }
    }
}

File: support/src/main/java/com/wuguangxin/ui/XinBaseActivity.kt
package com.wuguangxin.ui

import android.app.Activity
import android.app.Dialog
import android.content.Context
import android.content.Intent
import android.graphics.Color
import android.os.Build
import android.os.Bundle
import android.view.ViewGroup
import android.view.WindowManager
import androidx.annotation.ColorInt
import androidx.annotation.IdRes
import androidx.appcompat.app.AppCompatActivity
import androidx.databinding.DataBindingUtil
import androidx.databinding.ViewDataBinding
import com.wuguangxin.base.FragmentTask
import com.wuguangxin.base.LoadingStatus
import com.wuguangxin.dialog.LoadingDialog
import com.wuguangxin.dialog.XinDialog
import com.wuguangxin.listener.BaseInterface
import com.wuguangxin.utils.*
import com.wuguangxin.view.titlebar.TitleLayout

/**
 * Activity基类
 * Created by wuguangxin on 2015/4/1
 */
abstract class XinBaseActivity<B : ViewDataBinding> : AppCompatActivity(), BaseInterface {
    protected var titleLayout: TitleLayout? = null

    open var TAG: String? = ""
    open var loadingDialog: LoadingDialog? = null
    open var mDialog: XinDialog? = null
    lateinit var binding: B

    protected var screenWidth = 0
    private var slidingFinish = false
    private var mSlidingFinishHelper: SlidingFinishHelper? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = DataBindingUtil.setContentView(this, getLayoutId())
        loadingDialog = LoadingDialog(this) // 加载对话框
        mSlidingFinishHelper = SlidingFinishHelper(this)
        TAG = this::javaClass.name

        //SoftHideKeyBoardUtil.assistActivity(this)

        initArguments(intent)
        initView(savedInstanceState)
        setImmersionStatusBar();
        initListener()
    }

    /**
     * 设置沉浸式状态栏
     *
     * @param activity
     */
    fun setImmersionStatusBar(view: ViewGroup? = null) {
        val titleBar = view ?: titleLayout
        val barColor = titleLayout?.normalTheme?.backgroundColor?: Color.TRANSPARENT
        val isDark = titleLayout?.normalTheme?.isStatusBarDark?: true
        setImmersionStatusBar(this, titleBar, barColor, isDark)
        StatusBarUtils.setStatusBarFontColor(this, isDark)
    }

    open fun setImmersionStatusBar(
        activity: Activity,
        titleBarLayout: ViewGroup?,
        @ColorInt statusBarColor: Int,
        isDarkText: Boolean
    ) {
        if (titleBarLayout == null) {
            return
        }
        // >=19
        if (Build.VERSION.SDK_INT >= 19) {
            // 透明状态栏
            activity.window.addFlags(WindowManager.LayoutParams.FLAG_TRANSLUCENT_STATUS)
            // 透明导航栏。注意华为和HTC等有虚拟HOME键盘的，如果不设置下面这段代码，虚拟键盘将覆盖APP底部界面，无法操作底部TAB
            activity.window.addFlags(WindowManager.LayoutParams.FLAG_TRANSLUCENT_NAVIGATION)
            if (Build.VERSION.SDK_INT >= 21) {
                activity.window.addFlags(WindowManager.LayoutParams.FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS)
                activity.window.statusBarColor = statusBarColor
                titleBarLayout.setPadding(0, StatusBarUtils.getStatusBarHeight(activity), 0, 0)
            } else {
                /* <21
                android:fitsSystemWindows=""，
                false：布局不受StatusBar的影响，可以完全的展示在StatusBar的下面。
                true：布局不受StatusBar的影响，不会被StatusBar遮住，
                android:clipToPadding="false"
                false：布局不受Padding的影响，可以展示在Padding的区域。其实fitsSystemWindows就是设置一个Padding使View不会展示在StatusBar的下方，
                设置系统是否需要考虑 StatusBar 占据的区域来显示
                */
                titleBarLayout.fitsSystemWindows = false
                titleBarLayout.clipToPadding = true
            }
        }
        StatusBarUtils.setStatusBarFontColor(activity, isDarkText)
    }

    /**
     * 取消透明状态栏样式
     */
    fun clearImmersionStatusBar() {
        AndroidUtils.clearImmersionStatusBar(this)
    }

    /**
     * 是否开启滑动关闭 Activity
     */
    fun setSlidingFinish(slidingFinish: Boolean) {
        this.slidingFinish = slidingFinish
    }

    /**
     * 初始化传参
     */
    fun initArguments(intent: Intent) {
    }

    fun getContext(): Context = this

    override fun onStart() {
        super.onStart()
        initData()
    }

    override fun onDestroy() {
        super.onDestroy()
        dismissDialog()
    }

    override fun finish() {
        super.finish()
        AnimUtil.animClose(this)
    }

    open fun setTitleLayout(@IdRes id: Int): TitleLayout {
        val view = binding.root.findViewById<TitleLayout>(id).bind(this)
        this.titleLayout = view
        return view
    }

    override fun setTitle(title: CharSequence?) {
        super.setTitle(title)
        titleLayout?.title = title
    }

    override fun setTitle(titleId: Int) {
        super.setTitle(titleId)
        titleLayout?.setTitle(titleId)
    }

    override fun log(text: String) {
        Logger.i(TAG, text)
    }

    fun loge(text: String) {
        Logger.e(TAG, text)
    }

    override fun showToast(text: String) {
        ToastUtils.showToast(getContext(), text)
    }

    override fun openActivity(clazz: Class<out Activity>) {
        openActivity(clazz, null)
    }

    override fun openActivity(clazz: Class<out Activity>, bundle: Bundle?) {
        val intent = Intent(getContext(), clazz)
        intent.putExtra("bundle", bundle)
        startActivity(intent)
    }

    override fun startActivity(intent: Intent) {
        // 换成startActivityForResult打开的Activity切换动画比较好看，并且requestCode=0
        super.startActivityForResult(intent, 0)
    }

    fun setLoadingStatus(loadingStatus: Int, isPull: Boolean, isCached: Boolean) {
        if (loadingStatus == LoadingStatus.START) {
            if (isPull && isCached) {
                loadingDialog?.setVisible(true)
            }
        } else {
            loadingDialog?.setVisible(false)
        }
    }

    fun dismissDialog(vararg dialogs: Dialog?) {
        loadingDialog?.dismiss()
        loadingDialog = null
        mDialog?.dismiss()
        mDialog = null
        dialogs.forEach { it?.dismiss() }
    }

    public override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        // 让最后打开的那个fragment执行
        FragmentTask.topTask?.onActivityResult(requestCode, resultCode, data)
    }
}